linode/manager
# upcoming, change: [M3-7524] - Add child_account OAuth scope to Create and View PAT drawers

## Description 📝
New Scope:
Introduce a new OAuth scope named "child_account" for all users, but visible only to "Parent" accounts. As per API spec: its purpose is to allow Parent Account Users with the `child_account_access` grant to specify a Token with less permission than their User, i.e. a Token that is unable to list and/or create Tokens for the Child Account(s).

Token Configuration:
Default all scopes to `none` for users. Users should enable scopes as needed, contrary to having the initial state set to `read_write`. 

## Changes  🔄
- In the Create PAT drawer, a new "Child Account Access" scope is visible for **parent** users.
- In the View PAT drawer, a new "Child Account Access" scope is visible for **parent** users.

## Preview 📷

| Before  | After   |
| ------- | ------- |
| ![Screenshot 2023-12-13 at 4 01 40 PM](https://github.com/linode/manager/assets/114685994/690a6749-904c-4eba-a2ff-70ba2e03de27) | ![Screenshot 2023-12-13 at 4 00 48 PM](https://github.com/linode/manager/assets/114685994/33e3d5cd-4291-45eb-98e2-b329b97a94de) |
| ![Screenshot 2023-12-13 at 4 02 58 PM](https://github.com/linode/manager/assets/114685994/093b6d55-6f52-4d57-8c78-79c3b6f04baa) | ![Screenshot 2023-12-13 at 4 03 14 PM](https://github.com/linode/manager/assets/114685994/4e5f4e70-5554-42ea-aca0-98cf1a9a877d) |

## How to test 🧪

### Prerequisites
(How to setup test environment)
- Check out this PR and `yarn dev`.
- In the dev tools:
   - Make sure the Parent/Child Account Switching feature flag is on.
   - Turn mocks on.

### Verification steps 
(How to verify changes)
- Go to http://localhost:3000/profile/tokens and click the Create a Personal Access Token button.
- Observe that "Child Account Access" is visible as a new scope and all scopes default to None.
- On the tokens landing page, click the View Scope button for any mock PAT.
- Observe that the "Child Account Access" scope is visible in the table.
- Toggle the feature flag off and observe that the Child Account Access token is not visible in the Create or View drawers and all scopes default to Read/Write.
- Toggle the feature flag back on.
- Go to `serverHandler.ts` and edit the following request to change the user_type to `null`, `child`, or `proxy`:
```
  rest.get('*/account/users/:user', (req, res, ctx) => {
    // Parent/Child: switch the `user_type` depending on what account view you need to mock.
    return res(ctx.json(accountUserFactory.build({ user_type: 'parent' })));
  }),
```
- Observe that the Child Account Access token is not visible in the Create or View drawers.
- Verify tests pass:
```
yarn test utils CreateAPITokenDrawer ViewAPITokenDrawer

```

## As an Author I have considered 🤔

*Check all that apply*
